### DISCUSSION TITLE
POCL with custom device ?

### DISCUSSION THREAD

**Discussion by Alexix24** [pocl/pocl#1333] [General]
Hi,
I'm working on a project that aims to run OpenCL applications with a specific device : FPGA board that implements RiscV cores. The board reference is Terasic de5a-net. I have to create my own OpenCL driver/implementation and I can't use the Intel FPGA OpenCL driver that already exists.

Can I achieve this with Pocl ? I don't know how Pocl interacts with the device specific driver ?

**Edit:** I have found that Pocl calls some functions provided by the driver such as _compile\_kernel()_, _flush()_, _alloc\_mem\_obj()_ ... Where can I find the list of functions that the driver must implement ?

Thanks !

**Comment by pjaaskel** [comment]
@TopiLeppanen perhaps AlmaIF could be tested here?

**Comment by TopiLeppanen** [comment]
Hello, yes, your project sounds exactly the kind of thing I'm currently working on. I have been using custom devices and built-in kernels on various FPGAs. The cross-vendor driver I've been working on is in lib/CL/devices/almaif

More info and links to the publications can be found from the documentation http://portablecl.org/docs/html/almaif.html

I have so far tested this with Xilinx Pynq-Z1, Alveo U280 and Intel Arria 10 GX development kit. Unfortunately the Intel experiments aren't open-sourced, since I used some licensed components to achieve it.

Basically, I synthesize these AlmaIF devices on the FPGA, and then access the accelerator memory-map using the vendor-provided drivers. The PoCL's almaif driver then knows how to communicate with the AlmaIF-devices, so I get the cross-vendor FPGA portability quite easily.

**Reply by Shamanthbhargav** [reply]
Do we need to modify the cross vendor driver to support our custom device....and may i know the status of your work?

**Reply by TopiLeppanen** [reply]
Hi, yes, so depending on the hardware platform, you might need to modify the almaif cross vendor driver. I think currently open-sourced we have the support for XRT controlled devices and devices directly available in the physical memory map. Here's an illustration that I recently drew.
[almaif_driver_toolflow.pdf](https://github.com/pocl/pocl/files/14977849/driver_toolflow.drawio.pdf)
Out of these, the Altera and XDMA backends aren't yet in the open-source, but I'll likely upstream the XDMA backend soon

As a status of my own work, we recently published a conference paper in NorCAS conference that you can read about here. In that work we specified an automated FPGA bitstream management tool that can upload the bitstream to FPGA when user launches the kernel:
https://doi.org/10.1109/NorCAS58970.2023.10305457
And another even more recent work about OpenCL Pipe for FPGAs was recently presented in the IWOCL workshop:
https://dl.acm.org/doi/10.1145/3648115.3648128

Both of these works used the AlmaIF driver here in PoCL as the OpenCL runtime driver. The source code we used to generate the FPGA bitstreams is in: https://github.com/cpc/AFOCL

Apologies for the late answer, pinging me might be required, as that way I get the notification to my email.

**Comment by Lixiang9716** [comment]
I have the same problem. Actually , I want to use pocl to control zynq7045 like Alveo U280 , which is connected to a arm platform by pcie. Do you have any advices? @TopiLeppanen

**Reply by TopiLeppanen** [reply]
Hi, yes the AlmaIF driver can potentially be useful for you too, check my above reply. Do you have XRT or XDMA driver there to get access to the FPGA fabric?

**Reply by Lixiang9716** [reply]
Yes , I have installed   XDMA driver on the arm platform , and it runs well and  pass the basic  tests.
I think your work will greatly reduce the software and hardware co-design cycle, which will bring huge application value.
